### PR TITLE
Implement tagging system for song lists

### PR DESCRIPTION
**Pull request by ycanardeau**
#124

### CODE REVIEW

**Comment by riipah** [comment]
Thanks! I'll try to review tonight.

File: VocaDbWeb/Scripts/Repositories/SongListRepository.ts [hunk 1]
```diff
@@ -72,6 +78,10 @@ module vdb.repositories {
 
 		};
 
+		public getTagSuggestions = (songId: number, callback: (contract: dc.tags.TagUsageForApiContract[]) => void) => {
+			$.getJSON(this.urlMapper.mapRelative("/api/songLists/" + songId + "/tagSuggestions"), callback);
```
Comment (riipah) [inline VocaDbWeb/Scripts/Repositories/SongListRepository.ts]: I didn't see this method in SongListApiController. Are you intending to add tag suggestions based on songs in the list?

File: VocaDbModel/Database/Queries/UserQueries.cs [hunk 2]
```diff
@@ -1096,6 +1098,10 @@ class CachedUserStats {
 			return GetTagSelections<ReleaseEventSeries, EventSeriesTagUsage, EventSeriesTagVote>(seriesId, userId);
 		}
 
+		public TagSelectionContract[] GetSongListTagSelection(int songListId, int userId) {
```
Comment (riipah) [inline VocaDbModel/Database/Queries/UserQueries.cs]: Maybe GetSongListTagSelections

File: VocaDbWeb/Scripts/ViewModels/SongList/FeaturedSongListsViewModel.ts [hunk 3]
```diff
@@ -1,40 +1,44 @@
...
+				this.categories[categoryName] = new FeaturedSongListCategoryViewModel(listRepo, resourceRepo, tagRepo, languageSelection, cultureCode, categoryName);
 			});
 
-			this.initCategory(categoryNames[0]);
-
 		}
 
 		public categories: { [index: string]: FeaturedSongListCategoryViewModel; } = {};
 
-		public initCategory = (category: string) => {
```
Comment (riipah) [inline VocaDbWeb/Scripts/ViewModels/SongList/FeaturedSongListsViewModel.ts]: I couldn't figure out based on the code why this method is not needed. Can you explain?

File: VocaDbWeb/Views/Event/Details.cshtml [hunk 4]
```diff
@@ -232,7 +232,8 @@
 		$("#manageTags").button({ icons: { primary: 'ui-icon-wrench' } });
 
 		var loggedUserId = @UserContext.LoggedUserId;
-		var urlMapper = new vdb.UrlMapper('@UrlMapper.HostAddress');
+		var rootPath = '/';
```
Comment (riipah) [inline VocaDbWeb/Views/Event/Details.cshtml]: Hmm why did you hardcore rootpath to '/' instead of using the config parameter?

File: VocaDbWeb/Views/Event/SeriesDetails.cshtml [hunk 5]
```diff
@@ -97,7 +97,8 @@
 		$("#viewVersions").button({ icons: { primary: 'ui-icon-clock' } });
 
 		var loggedUserId = @UserContext.LoggedUserId;
-		var urlMapper = new vdb.UrlMapper('@UrlMapper.HostAddress');
+		var rootPath = '/';
```
Comment (riipah) [inline VocaDbWeb/Views/Event/SeriesDetails.cshtml]: Here too

File: VocaDbWeb/Views/Shared/Partials/_SongListsFilters.cshtml [hunk 6]
```diff
@@ -1,22 +1,43 @@
...
-			</a>
-			<a data-bind="click: function() { sort('CreateDate'); }, css: { active: sort() == 'CreateDate' }" class="btn" href="#">
-				@VocaDb.Web.Resources.Views.SongList.SongListSortRuleNames.CreateDate
-			</a>
-			<a data-bind="click: function() { sort('Name'); }, css: { active: sort() == 'Name' }" class="btn" href="#">
-				@VocaDb.Web.Resources.Views.SongList.SongListSortRuleNames.Name
-			</a>
+﻿<div class="form-horizontal well well-transparent" data-bind="visible: true" style="display: none;">
+	
+	<div class="pull-right">
```
Comment (riipah) [inline VocaDbWeb/Views/Shared/Partials/_SongListsFilters.cshtml]: I don't know why did you change this to dropdown instead of tabs, but I guess it works like this too.

File: VocaDbWeb/Views/SongList/Featured.cshtml [hunk 7]
```diff
@@ -52,10 +52,19 @@
...
 
-		var urlMapper = new vdb.UrlMapper('@RootPath');
+		var languageSelection = '@UserContext.LanguagePreference';
+		var rootPath = '@RootPath';
+		var urlMapper = new vdb.UrlMapper(rootPath);
+		var repoFactory = new vdb.repositories.RepositoryFactory(urlMapper, @LanguagePreferenceInt, @UserContext.LoggedUserId);
+		var songListRepo = repoFactory.songListRepository();
+		var resourceRepo = repoFactory.resourceRepository();
+		var tagRepo = repoFactory.tagRepository();
+		var languageSelection = '';
```
Comment (riipah) [inline VocaDbWeb/Views/SongList/Featured.cshtml]: Why are you overriding languageSelection here?

**Comment by riipah** [comment]
I think TagTargetTypes should be updated, even if you're not doing tag filtering by entry type (that can be a separate task).

**Comment by riipah** [comment]
TagDetailsContract should be updated. At least the number of song lists.

File: VocaDbModel/Domain/Tags/Tag.cs [hunk 8]
```diff
@@ -401,6 +402,14 @@ public Tag(LocalizedString name, string categoryName = "")
 			}
 		}
 
+		public virtual ISet<SongListTagUsage> AllSongListTagUsages {
```
Comment (riipah) [inline VocaDbModel/Domain/Tags/Tag.cs]: This should be added to AllTagUsages. Events too (I forgot them before).

**Comment by riipah** [comment]
TagUsageQueries.GetActualTagUsageCounts(...) should be updated too, so that it takes into account song list tag usage counts.

**Comment by riipah** [comment]
Thanks for the PR! I don't blame you for missing those few things, unfortunately there's quite a lot of places that you just need to know when adding a new tag type.

File: VocaDbWeb/Views/SongList/Featured.cshtml (hunk 7 above)
Comment (ycanardeau) [reply VocaDbWeb/Views/SongList/Featured.cshtml]: I guess it must be a mistake. Thanks.

File: VocaDbWeb/Views/Shared/Partials/_SongListsFilters.cshtml (hunk 6 above)
Comment (ycanardeau) [reply VocaDbWeb/Views/Shared/Partials/_SongListsFilters.cshtml]: Yes, and I think this is more consistent with other layouts.

File: VocaDbWeb/Views/Event/Details.cshtml (hunk 4 above)
Comment (ycanardeau) [reply VocaDbWeb/Views/Event/Details.cshtml]: Fixed, but @UrlMapper.HostAddress should not be used because it generates a non-SSL URL (e.g. http://localhost:39390/ instead of https://localhost:44399/).

File: VocaDbWeb/Scripts/ViewModels/SongList/FeaturedSongListsViewModel.ts (hunk 3 above)
Comment (ycanardeau) [reply VocaDbWeb/Scripts/ViewModels/SongList/FeaturedSongListsViewModel.ts]: This method is not needed because the clear method gets called when the getList method succeeded. If this method is present, the loadMore method gets called twice when first loading and duplicate song lists appear.

**Comment by ycanardeau** [comment]
Thank you for your review!

**Comment by ycanardeau** [comment]
I updated TagDetailsContract, but I haven't changed the view. By the way, the "Show more" button for event series is linked to the event search page. Is this correct?

**Comment by riipah** [comment]
> I updated TagDetailsContract, but I haven't changed the view. 

Maybe you could add the number of song lists at least? And link to featured song lists (even though it's not all).

>By the way, the "Show more" button for event series is linked to the event search page. Is this correct?

Yeah that's intentional. For now event search page is also good enough for series.

**Comment by ycanardeau** [comment]
Do you mean like this? Or should I add "Top song lists" to the view? I don't think "All song lists" is an appropriate name though because it is actually not "all"...

**Comment by riipah** [comment]
Should be ok, thanks!
